src/Language/PureScript/Sugar/CaseDeclarations.hs

```diff
@@ -39,6 +38,186 @@ desugarCasesModule (Module ss coms name ds exps) =
       <*> pure exps
 
 -- |
+-- Desugar case with pattern guards and pattern clauses to a
+-- series of nested case expressions.
+--
+desugarCase :: (Monad m, MonadSupply m)
+            => Expr
+            -> m Expr
+desugarCase (Case scrut alternatives) =
+  let
+    -- Alternatives which do not have guards are
+    -- left as-is. Alternatives which
+    --
+    --   1) have multiple clauses of the form
+    --      binder | g_1
+    --             , g_2
+    --             , ...
+    --             , g_n
+    --             -> expr
+    --
+    --   2) and/or contain pattern guards of the form
+    --      binder | pat_bind <- e
+    --             , ...
+    --
+    -- are desugared to a sequence of nested case expressions.
+    --
+    -- Consider an example case expression:
+    --
+    --   case e of
+    --    (T s) | Just info <- Map.lookup s names
+    --          , is_used info
+    --          -> f info
+    --
+    -- We desugar this to
+    --
+    --   case e of
+    --    (T s) -> case Map.lookup s names of
+    --               Just info -> case is_used info of
+    --                              True -> f info
+    --                              (_    -> <partial>)
+    --               (_ -> <partial>)
+    --
+    -- Note that if the original case is partial the desugared
+    -- case is also partial.
+    --
+    -- Consider an exhaustive case expression:
+    --
+    --   case e of
+    --    (T s) | Just info <- Map.lookup s names
+    --          , is_used info
+    --          -> f info
+    --    _     -> Nothing
+    --
+    -- desugars to:
+    --
+    --    case e of
+    --      _ -> let
+    --                v _ = Nothing
+    --           in
+    --             case e of
+    --                (T s) -> case Map.lookup s names of
+    --                          Just info -> f info
+    --                          _ -> v true
+    --                _  -> v true
+    --
+    -- This might look strange but simplifies the algorithm a lot.
+    --
+    dsAlt :: (Monad m, MonadSupply m) => [CaseAlternative] -> m [CaseAlternative]
+    dsAlt [] = pure []
+
+    -- the trivial case: no guards
+    dsAlt (a@(CaseAlternative _ [([], _)]) : as) =
+      (a :) <$> dsAlt as
+
+    -- we make sure to clean up all single conditional clauses
+    -- as we would otherwise generate the out-of-line code for
+    -- them which is unecessary.
+    dsAlt (CaseAlternative ab ge : as)
+      | not (null cond_guards) =
+          (CaseAlternative ab cond_guards :)
+            <$> dsGrdAlt ab rest as
+      | otherwise = dsGrdAlt ab ge as
+      where
+        (cond_guards, rest) = span isSingleCondGuard ge
+
+        isSingleCondGuard ([ConditionGuard _], _) = True
```

Comment:
Good point. I thought about this too and since we can express conditional guards in CoreFn natively I decided to keep. If we decide not to special case it I think support for guards can be removed in CoreFn.